### ISSUE TITLE
CPCSS for specific pages: UI & metabox (4/6)

### ISSUE THREAD

**Issue by Tabrisrp** [wp-media/wp-rocket#2443]
Sub-Card for epic #2377.

This card is for tasks related to the metabox to display on the `edit.php` screen.

**Summary of the following discussion for the UI**: https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-617579591

The following was the contents of the original post: 

```
Currently we need to have a final decision made on:
- where and how to display the buttons to (re-)generate/delete the specific CPCSS for the current page
- The wording for the buttons
- The success/error messages after generation/deletion

What was proposed:
### Add the button to our existing metabox
**Pros**
- Everything WP Rocket is grouped in the same place

**Cons**
- The button looks too much like it needs to be "clicked" to validate any change inside the metabox

### Add a new metabox
**Pros**
- Clear separation of concerns

**Cons**
- Content added by WP Rocket is split in different places

Depending on the decision, I'll update this accordingly to determine the development tasks to do for this part.

@arunbasillal @GeekPress @webtrainingwheels 
```

**Comment by Tabrisrp** [comment]
@hellofromtonya tagging you on this one too

**Comment by GeekPress** [comment]
@arunbasillal What is your opinion about the solution to adopt?

**Comment by arunbasillal** [comment]
Another approach to the UI

What if we follow the rest of the options and make it into an entry in the existing metabox? 

<img width="205" alt="custom CPCSS" src="https://user-images.githubusercontent.com/5751474/77214100-68901a80-6b33-11ea-96cf-aa0796955522.png">

- The checkbox is disabled by default for all posts / pages when the customer updates to 3.6.
- If Optimize CSS delivery is disabled, the checkbox is greyed out. 
- When a user wants custom CPCSS for the specific page/post, they check the box and save the post. As with every other option. 
- To switch back to the default CPCSS, the user unchecks the `Generate custom CPCSS` checkbox and updates the post. Just like every other option. 

So far so good I hope. The challenge now is the regeneration of CPCSS. 

- If we can find a way to tie this into the global `Regenerate CPCSS` that would solve the challenge. 
- If you think about it, if the customer changes the theme and wants to regenerate the CPCSS for the entire website, they shouldn't have to go into individual posts/pages to do so. 
- We could save the details about the post / pages that uses custom CPCSS in our database (when the customer enables it) and regenerate all of them after the default ones.

**Comment by Tabrisrp** [comment]
This card is in sprint 4, but the UI is not finalized. Discussion needs to start again before we start dev on this part.

Personal opinion: I don't like the checkbox. It doesn't feel natural, and is more complex than using buttons.

**Comment by GeekPress** [comment]
I'm a fan of the checkbox too, and I wasn't to find somehting than a new metabox:
![Default Metabox](https://share.getcloudapp.com/GGukXrjP)

And the state when we have a custom CPCSS
![Metabox with custom CPCSS](https://share.getcloudapp.com/jkul7DNb)

@arunbasillal @webtrainingwheels Can you please work on the working? :)

**Comment by arunbasillal** [comment]
Some feedback regarding the `Generate Custom CPCSS` button for discussion:

- What happens when a user clicks the button? We send an ajax request to generate the CPCSS? Or reload the page? 
- If we reload the page, we have to save the users edits before we do that. So that isn't ideal and we need to use ajax I think. 

Another thought: 
- User clicks the button. The post is not published yet. CPCSS generation will fail.
- Customer publishes the post. There is no custom CPCSS. 
- WP Rocket has to somehow "remember" user's click and generate the CPCSS when the post is published. 

Using a checkbox will solve the above two problems. 

We need a separate section, otherwise the user will have no idea what the option means. However, I am still leaning towards the checkbox but with it's own dedicated section. 

Here is a proposal

Default state: 
![image](https://user-images.githubusercontent.com/5751474/79581514-22a18480-80e8-11ea-9c02-82aafec5c7eb.png)

Posts using custom CPCSS:
![image](https://user-images.githubusercontent.com/5751474/79581942-bbd09b00-80e8-11ea-9962-653ebc818af8.png)

**Comment by GeekPress** [comment]
I'm not against avoiding a new metabox and add the Custom CPCSS into our actual one. It seems more intuitive like this in fact ^^

But I'm still not a fan if using a checkbox here. The label « Use » a quite weird ^^ What does it mean?

When a button with the « Generate Custom CPCSS » is less confused and don't need to save the options to validate the choice.

**Comment by arunbasillal** [comment]
@GeekPress 

> But I'm still not a fan if using a checkbox here. The label « Use » a quite weird ^^ What does it mean?

The word will be `Generate` when there is no custom CPCSS. Then changed to `Use`. It goes with the rest of the wording in the meta box. Don't you think? (The wording is different when a post using custom CPCSS)

>When a button with the « Generate Custom CPCSS » is less confused and don't need to save the options to validate the choice.

This could lead to the second issue mentioned here: https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-615287757

**Comment by Tabrisrp** [comment]
> What happens when a user clicks the button? We send an ajax request to generate the CPCSS? Or reload the page?

It's an AJAX request.

> User clicks the button. The post is not published yet. CPCSS generation will fail.

The button will be disabled if the post status is not published.

I'm strongly against the checkbox, it doesn't feel natural at all.

**Comment by webtrainingwheels** [comment]
@Tabrisrp @GeekPress What makes you against the checkbox?
Technically-speaking, is it an incorrect use of a checkbox? 
Or is it more your personal preferences?

A separate metabox with a bold blue button feels like it emphasizes this option too much and places more importance on it than it should have. From that perspective the checkbox in a separate area of the same metabox, feels less obtrusive. But if it's a bad practice to use a checkbox as a trigger to generate something, I respect that. We don't want to be janky and hacky.

If we do go with the separate metabox, is there a more subtle button style we can go with? It seems like the ability to generate custom cpcss shouldn't have the same visual importance as the default WordPress Update/Publish button.

**Comment by GeekPress** [comment]
@webtrainingwheels A checkbox should be used to activate / deactivate an option. Here, we are requesting an action: generate a custom CPCSS, it's not an « option » like all other are.

And more important, we need to have a response of the action to validate the action: success / failure. It won't be possible with a checkbox. We will tick it, save the option, and it won't be possible to know if the action was done. Oh yes, we could... with a new notice :D

**Comment by Tabrisrp** [comment]
> If we do go with the separate metabox, is there a more subtle button style we can go with? It seems like the ability to generate custom cpcss shouldn't have the same visual importance as the default WordPress Update/Publish button.

Yes we can use the secondary style of course, same as the example for the "regenerate" button

**Comment by GeekPress** [comment]
Ok, let's do a mix of the suggestions. We will use the same metabox to display the « Critical Path CSS » section.

And we will use buttons to generate, re-generate or cancel the custom CPCSS like here:
![Default Metabox](https://p195.p4.n0.cdn.getcloudapp.com/items/GGukXrjP/Capture%20d%E2%80%99e%CC%81cran%202020-04-14%20a%CC%80%2011.34.00.png?v=09b2410122edd88ca83dc17f01cafb9e)

And the state when we have a custom CPCSS
![Metabox with custom CPCSS](https://p195.p4.n0.cdn.getcloudapp.com/items/jkul7DNb/Capture%20d%E2%80%99e%CC%81cran%202020-04-14%20a%CC%80%2011.38.45.png?v=ac6d95ebe8242cd2ed369c7f611c6380)

**Comment by Tabrisrp** [comment]
To note: adding it to the existing metabox will increase development time, because we will have to move the code to the new architecture, and create tests for the whole existing metabox content + callbacks, in addition to the new part.

**Comment by hellofromtonya** [comment]
@Tabrisrp The increase in dev time is okay as it provides us with more code coverage and shifts more code in our new architecture. 

Please update the effort estimate. Thanks.

**Comment by arunbasillal** [comment]
Since we have come to a collective decision, I am trying to summarize our approach and finalize some finer details. 

- We will be using the **existing metabox**.
- All actions happens over AJAX.
- Sub-section title: **Custom CPCSS**
- Sub-section description: **Generate custom CPCSS for this post. More info** (before generation) and **This post uses custom CPCSS. More info** (after generation). 
- Button text will say **Generate Custom CPCSS** before a custom CPCSS is generated for the post
- After a custom CPCSS is generated for the post, it will say **Regenerate Custom CPCSS** (No hyphen needed, we do not use hyphen in Admin bar - https://jmp.sh/shU5dYB)
- The button style will be **secondary** before and after custom CPCSS is generated for the post (https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-617118954)
- The button will stay disabled under two conditions: **post status is not published**, **Optimize CSS delivery is disabled** in WP Rocket settings (https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-615396810) (We can inform customer about this via a small text under the button.)
- Text when button is disabled ^: **Enable Optimize CSS delivery in WP Rocket settings and publish the post to use this feature** (only display relevant reason if possible). 
- When a post uses Custom CPCSS, the button will have a link under it that reads **or revert back to the original CPCSS**
- On clicking this ^ link, the custom CPCSS is deleted and the button resets to the default state (i.e. button will now read `Generate Custom CPCSS` and the link will disappear). 

**Action when "Generate custom CPCSS" button is clicked:** 
- Disable the button temporarily with a spinner in it to show progress - https://jmp.sh/OoCSx3p
- On success, display: **Custom CPCSS generated successfully** under the button.
- On success, the button becomes the regenerate button with the link to cancel it underneath. 
- On error, display **Custom CPCSS generation failed. Error: Error message xxxx** under the button. 

**Other notes:** 
- When customer clicks `Regenerate Critical Path CSS` in the admin bar or WP Rocket dashboard, the custom CPCSS should also be regenerated. 
- Also when WP Rocket regenerates CPCSS when a theme is changed (and other instances), custom CPCSS should also be regenerated. 

@GeekPress @Tabrisrp  Please take a look and confirm.

**Comment by GeekPress** [comment]
@hellofromtonya Can we be a little flexible here about doing the code rewrite right now?

The new effort is a M size and we already have one sprint late. Spending a week on this will mandatory have an impact on the next sprints. I added one more sprint for 3.6 and we won't be able to add another one as we will have a promo campaign in June too.

If we can't be flexible, then please do a new metabox to keep the previous effort.

**Comment by GeekPress** [comment]
@arunbasillal Thanks a lot for the summarize! 

Here some suggestions:
Sub-section title: Custom CPCSS

We are using the term « Optimize CSS Delivery » in the checkbox list, and we are going to use another term. I'm not sure that people will totally understand the relation and what it's going to do.

On the sub-description, maybe we could be more specific.
> The button style will be secondary before and after custom CPCSS is generated for the post (#2443 (comment))

The buttons style can't be the same on both buttons as the actions are totally different. You're about not using the primary button (blue color) into a metabox. WordPress doesn't this style on metabox.

Here a new suggestion:
The « Generate Custom CPCSS » button uses the secondary style (no background + with bleu border).

« Regenerate Custom CPCSS » is a link with default style (blue underline).
« revert back to the original CPCSS » is a link with « is destructive style » like here:
https://share.getcloudapp.com/lluDyJD1

To have this style, the link needs to have these CSS classes: `.components-button .is-link .is-destructive`

**Comment by Tabrisrp** [comment]
> When customer clicks Regenerate Critical Path CSS in the admin bar or WP Rocket dashboard, the custom CPCSS should also be regenerated.
    Also when WP Rocket regenerates CPCSS when a theme is changed (and other instances), custom CPCSS should also be regenerated.

Was this ever mentioned in the original doc? It has not been planned for, current development is not performing any auto-update of the posts CPCSS.

**Comment by arunbasillal** [comment]
@Tabrisrp 

> Was this ever mentioned in the original doc? It has not been planned for, current development is not performing any auto-update of the posts CPCSS

I believe not. Sorry. This came out of this comment: https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-601958310

How hard is it to add it to current sprint? If it's too much, we could add it as an enhancement in the future I hope. We can open an issue for now so that we do not forget it.

**Comment by Tabrisrp** [comment]
It would at least add 1-2 days of dev + QA time. We would also have to discuss about how to best store which posts have a specific CPCSS generated.

**Comment by arunbasillal** [comment]
Thanks @Tabrisrp. I believe it's @GeekPress's call on which sprint to add it on. 

I have updated my summary (https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-617579591) with the feedback from Jonathan (https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-617622062)

**Comment by Tabrisrp** [comment]
Questions:
- Where should the _more info_ link point to? I guess a documentation link?
- Does "_custom_" makes the most sense here? Isn't "_specific_" a more accurate description of what this is?
- What should we do when specific CPCSS has been created for one or more posts, and after that the Optimize CSS Delivery option is disabled? Or the post status is no longer _publish_?

> revert back to the original CPCSS

I feel "_original CPCSS_" might be confusing. Can the link text be changed to "_delete specific CPCSS for this post_" instead, or something similar?

**Comment by GeekPress** [comment]
> It would at least add 1-2 days of dev + QA time. We would also have to discuss about how to best store which posts have a specific CPCSS generated.

Will it be a new sub-task? If so, you can create and we will plan it for the next sprint.

> Where should the more info link point to? I guess a documentation link?

Yes you're right. We plan to have a documentation for this feature. It will be linked on this « More info ».

> Does "custom" makes the most sense here? Isn't "specific" a more accurate description of what this is?

Indeed, « specific » is more accurate than « custom ». 
@arunbasillal @webtrainingwheels Is it ok for you?

> What should we do when specific CPCSS has been created for one or more posts, and after that the Optimize CSS Delivery option is disabled?

If the Optimize CSS Delivery option is disabled, the specific CPCSS should be disabled too.

> Or the post status is no longer publish?
We can do the same behavior than we are doing now for the Optimize CSS Delivery option.

> I feel "original CPCSS" might be confusing. Can the link text be changed to "delete specific CPCSS for this post" instead, or something similar?

@Tabrisrp What about the a mix like: revert back to the default CPCSS for this post

I'm wondering if we use the term « delete » that our user could think that it could also « delete » the default CPCSS.

**Comment by Tabrisrp** [comment]
After discussing about the approach with Tonya, it was decided that:
- The CPCSS part would be added to the metabox via a `do_action()` call inside the metabox
- The current metabox code would not be changed as part of this PR
- The effort is back to `[S]` with this approach
- If/When we want to update the metabox code, preserving the `do_action()` call will be enough to keep the CPCSS part with no changes required

**Comment by GeekPress** [comment]
@Tabrisrp @hellofromtonya Thanks for the update, and also for being flexible 👏

**Comment by webtrainingwheels** [comment]
>Indeed, « specific » is more accurate than « custom ».
@GeekPress agree - "specific" is good for me.

**Comment by arunbasillal** [comment]
Specific seems better to me as well. I have updated the summary - https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-617579591

Changes: 

- Sub-section title changed to **Critical Path CSS** (was Custom CPCSS before. Specific CPCSS doesn't seem to fit) 
- Sub-section description: "custom" replaced with "specific" as discussed. 
- Button text will be: **Generate Specific CPCSS**
- Revert link will be: **or revert back to the default CPCSS** (shortened from "revert back to the default CPCSS for this post" to avoid a really long link)

**Comment by Tabrisrp** [comment]
Additional question: how should we handle classic vs new editor?

a good part of the styles available for buttons with the new editor, are not available when switching to classic, so the display isn't great (still useable of course).

Do we want to have the same display on both, or let it degrade on the classic?

I'm trying to avoid adding additional styles on the page just for backward design compatibility.

**Comment by arunbasillal** [comment]
@Tabrisrp @GeekPress For this question - https://github.com/wp-media/wp-rocket/issues/2443#issuecomment-619200123

I am fine with sticking to the new editor and letting it degrade in Classic. I have added a note for testing in the main card to see how it turns out in the end. 

Is there a way to know what percentage of our users use the new editor and vice versa?

**Comment by GeekPress** [comment]
@Tabrisrp Can we have a screenshot with the classic editor to be able to view the rendering?

**Comment by Tabrisrp** [comment]
Classic editor, generate button:
![classic-generate](https://user-images.githubusercontent.com/3465180/80617277-e63b2600-8a0f-11ea-83a8-501bb8ca5164.jpg)

Classic editor, regenerate/delete buttons:
![classic-regenerate](https://user-images.githubusercontent.com/3465180/80617281-e6d3bc80-8a0f-11ea-94a5-39223f1ee9d2.jpg)

**Comment by GeekPress** [comment]
@Tabrisrp Thanks. It's fine to keep like that for me.
